src/NuGet.Core/NuGet.Common/KeyedMutex.cs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Common
+{
+    internal sealed class KeyedMutex : IDisposable
+    {
+        private readonly Dictionary<string, LockState> _locks;
+        private readonly SemaphoreSlim _dictionaryLock;
+
+        internal KeyedMutex()
+        {
+            _locks = new Dictionary<string, LockState>();
+            _dictionaryLock = new SemaphoreSlim(initialCount: 1);
+        }
+
+        internal async Task EnterAsync(string key, CancellationToken token)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            LockState lockState;
+
+            await _dictionaryLock.WaitAsync(token);
```

Comment:
Just curious: why is a semaphore slim used here instead of `lock`? `GetOrCreate` is synchronous so an async-compatible locking mechanism does not seem necessary. Is there a perf difference?

Reply:
If Thread1 has the lock, and Thread2 tries to get the lock, using `SemaphoreSlim.WaitAsync`, it can return an incomplete task and the thread can do other work. With `lock`, Thread2 will block. This has the performance issues of the pre-async world.